### PR TITLE
Improvement/types

### PR DESCRIPTION
**Pull request by tmkn**
**Description:**
This PR improves once again the typings.
`strict` mode errors are down from 91 to 50 🚀

### CODE REVIEW

File: src/components/Login/Login.tsx [hunk 1]
```diff
@@ -131,22 +130,19 @@ export default class LoginModal extends Component<Partial<LoginModalProps>, Logi
 
   public submitCredentials = async () => {
     const { form } = this.state;
-    const username = (form.username && form.username.value) || '';
-    const password = (form.password && form.password.value) || '';
+    const username = form.username?.value ?? '';
```
Comment (tmkn) [inline src/components/Login/Login.tsx]: null coalescing operator and optional chaining at work 💪

File: src/components/Login/Login.tsx [hunk 2]
```diff
@@ -98,31 +100,28 @@ export default class LoginModal extends Component<Partial<LoginModalProps>, Logi
...
       {
-        form: Object.keys(form).reduce(
-          (acc, key) => ({
-            ...acc,
-            ...{ [key]: { ...form[key], pristine: false } },
-          }),
-          { username: {}, password: {} }
-        ),
+        form: {
+          username: { ...form.username, pristine: false },
```
Comment (tmkn) [inline src/components/Login/Login.tsx]: I opted for a simple assignment since the keys weren't dynamic and the reduce construct was hard to understand for TypeScript.

File: src/components/Login/Login.tsx [hunk 3]
```diff
@@ -131,22 +130,19 @@ export default class LoginModal extends Component<Partial<LoginModalProps>, Logi
 
   public submitCredentials = async () => {
     const { form } = this.state;
-    const username = (form.username && form.username.value) || '';
-    const password = (form.password && form.password.value) || '';
+    const username = form.username?.value ?? '';
+    const password = form.password?.value ?? '';
     const { onSubmit } = this.props;
     if (onSubmit) {
       await onSubmit(username, password);
```
Comment (tmkn) [inline src/components/Login/Login.tsx]: While not an error, vscode suggested to remove the unnecessary `await`.
Turns out `onSubmit` was typed with `void` and not `Promise<void>` for its return value, fixed that as well
https://github.com/verdaccio/ui/blob/00d6223ad85f753c31e21e8eb3e85120d665f38a/src/components/Login/Login.tsx#L33

File: src/components/Login/Login.tsx (hunk 1 above)
Comment (juanpicado) [reply src/components/Login/Login.tsx]: nice !! glad you can use it already !

File: src/components/Login/Login.tsx (hunk 3 above)
Comment (priscilawebdev) [reply src/components/Login/Login.tsx]: well done.

File: src/components/RegistryInfoContent/RegistryInfoContent.tsx [hunk 4]
```diff
@@ -66,7 +66,7 @@ const RegistryInfoContent: React.FC<Props> = props => {
   };
 
   /* eslint react/prop-types:0 */
-  const TabContainer = ({ children }): JSX.Element => {
+  const TabContainer: React.FC = ({ children }) => {
```
Comment (priscilawebdev) [inline src/components/RegistryInfoContent/RegistryInfoContent.tsx]: 👏

File: src/utils/test-enzyme.ts [hunk 5]
```diff
@@ -4,13 +4,13 @@ import ThemeProvider from '../design-tokens/ThemeProvider';
...
 
-const shallowWithTheme = (element: React.ReactElement<any>, ...props): any =>
+const shallowWithTheme = (element: React.ReactElement<any>, ...props: undefined[]): any =>
   shallow(element, {
     wrappingComponent: ThemeProvider,
     ...props,
   });
 
-const mountWithTheme = (element: React.ReactElement<any>, ...props): any =>
+const mountWithTheme = (element: React.ReactElement<any>, ...props: undefined[]): any =>
```
Comment (priscilawebdev) [inline src/utils/test-enzyme.ts]: Sorry, but I am not sure I agree with this type 🧐

File: src/components/Search/Search.tsx [hunk 6]
```diff
@@ -25,7 +25,7 @@ export type handleClickSearch = (
   event: KeyboardEvent<HTMLInputElement>,
   { suggestionValue, method }: { suggestionValue: object[]; method: string }
 ) => void;
-export type handleFetchPackages = ({ value: string }) => Promise<void>;
+export type handleFetchPackages = (arg: { value: string }) => Promise<void>;
```
Comment (priscilawebdev) [inline src/components/Search/Search.tsx]: Please, why do you find "arg:" necessary? I believe to be irrelevant

File: src/components/Search/Search.test.tsx [hunk 7]
```diff
@@ -246,7 +246,7 @@ describe('<Search /> component test', () => {
       const getDetailPageURL = jest.fn(() => 'detail/page/url');
       jest.doMock(URL_FILE_PATH, () => ({ getDetailPageURL }));
 
-      const suggestionValue = [];
+      const suggestionValue: unknown[] = [];
```
Comment (priscilawebdev) [inline src/components/Search/Search.test.tsx]: I would prefer to create an interface for the suggestionValue, but since I am refactoring the AutoComplete component we can leave it like that

File: src/components/Search/Search.test.tsx [hunk 8]
```diff
@@ -155,7 +155,7 @@ describe('<Search /> component test', () => {
     beforeEach(() => {
       jest.resetModules();
       jest.doMock('lodash/debounce', () => {
-        return function debounceMock(fn) {
+        return function debounceMock(fn: unknown) {
```
Comment (priscilawebdev) [inline src/components/Search/Search.test.tsx]: Please are you sure about this type?

File: src/components/Repository/Repository.test.tsx [hunk 9]
```diff
@@ -19,7 +20,7 @@ const mockPackageMeta: jest.Mock = jest.fn(() => ({
 }));
 
 jest.mock('../../pages/Version', () => ({
-  DetailContextConsumer: component => {
+  DetailContextConsumer: (component: { children: (arg0: { packageMeta: PackageMetaInterface }) => void }) => {
```
Comment (priscilawebdev) [inline src/components/Repository/Repository.test.tsx]: sorry, but I am not so sure I like it...maybe there is a simpler way?

File: src/components/Search/Search.tsx (hunk 6 above)
Comment (tmkn) [reply src/components/Search/Search.tsx]: Haha this one was tricky, I didn't encounter this before 😉 
but that particular syntax didn't work as intended.
tldr: It only **renamed** `value` to `string`.
It didn't say that `value` is of type `string`.
see https://github.com/microsoft/TypeScript/issues/9657

To interpret `{ value: string }` as a type like intended and not a rename in a destructure assignment (which is apparently possible) you need to add the argument name.

File: src/components/Search/Search.test.tsx (hunk 8 above)
Comment (tmkn) [reply src/components/Search/Search.test.tsx]: Since this is only a type to satisfy the unit test and called internally by jest aka never called manually by production code I just used the fastest route to satisfy the compiler.

File: src/components/Repository/Repository.test.tsx (hunk 9 above)
Comment (tmkn) [reply src/components/Repository/Repository.test.tsx]: It was auto generated by vscode, it can now generate types based on the usage 🚀 
Didn't think it would work but it worked.
I can properly type this if wanted.

File: src/utils/test-enzyme.ts (hunk 5 above)
Comment (tmkn) [reply src/utils/test-enzyme.ts]: Again since this is only used to type a unit test and not called by production code I took the fastest route to satisfy the compiler.

Comment (priscilawebdev) [reply src/utils/test-enzyme.ts]: Hm..I understand your point. I would do so

image

....but we have this lint rule

The thing is that if the type is "any" or the build is not satisfied, then we know that we need to right the correct type, but when we introduce, for instance undefined[], we might not see where we should fix in the future... do you understand? so even if it's not production code, I would appreciate to have it correct..otherwise leave as it is

File: src/components/Search/Search.tsx (hunk 6 above)
Comment (priscilawebdev) [reply src/components/Search/Search.tsx]: Very interesting! but in this case i would then do: `({value}: { value: string})`. It's more readable IMO, otherwise people can get confused

File: src/components/Search/Search.test.tsx (hunk 8 above)
Comment (priscilawebdev) [reply src/components/Search/Search.test.tsx]: I see...but maybe it's better to right a function type, since this is a function right? for example: `fn: () => void`. some investigation might be need to check what we could write to replace void..

Comment (priscilawebdev) [reply src/components/Search/Search.test.tsx]: Maybe just a `Function` type would be also fine

File: src/components/Repository/Repository.test.tsx (hunk 9 above)
Comment (priscilawebdev) [reply src/components/Repository/Repository.test.tsx]: Hm... I refactored this component today... I've removed this part of the code, so that is fine :-)

File: src/components/Login/Login.tsx (hunk 1 above)
Comment (DanielRuf) [reply src/components/Login/Login.tsx]: Do we transpile this? It is only availble in Node 13+ afaik.

File: src/components/Repository/Repository.test.tsx (hunk 9 above)
Comment (DanielRuf) [reply src/components/Repository/Repository.test.tsx]: Well but then we have to resolve the conflicts now ;-)

**Comment by priscilawebdev** [comment]
@tmkn please what is the status of this PR?

**Comment by tmkn** [comment]
@priscilawebdev will update this pr this weekend, just don't know whether its worth the effort to merge and resolve the conflicts or start with a fresh master since the type improvements where straight forward.

**Comment by juanpicado** [comment]
As you wish, I would do be same

**Comment by tmkn** [comment]
Will open a new PR for the improved typings so that it doesn't drag along all the accumulated baggage from this.

**Comment by priscilawebdev** [comment]
I would do the same. Thanks @tmkn :)
